### PR TITLE
Add Raftis Jepsen Test

### PR DESCRIPTION
**Pull request by gaodq**
Hi, Kyle
We have tested out Raft implementation [Floyd](https://github.com/Qihoo360/floyd) with Jepsen.

### CODE REVIEW

File: raftis/src/jepsen/raftis.clj [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+(def src_dir "/home/gaodunqiao/raftis")
+(def dir     "/opt/raftis")
+(def logfile (str dir "/data/LOG"))
+(def binary  "raftis")
+
+(defn db
+  "Raftis DB for a particular version."
+  [version]
+  (reify db/DB
+    (setup! [_ test node]
```
Comment (aphyr) [inline raftis/src/jepsen/raftis.clj]: It looks like you haven't provided a way to install the database; without code to do that, other users won't be able to run your test.

File: raftis/src/jepsen/raftis.clj [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    (teardown! [_ test])))
+
+(def src_dir "/home/gaodunqiao/raftis")
+(def dir     "/opt/raftis")
+(def logfile (str dir "/data/LOG"))
+(def binary  "raftis")
+
+(defn db
+  "Raftis DB for a particular version."
+  [version]
```
Comment (aphyr) [inline raftis/src/jepsen/raftis.clj]: You provide a version here, but it's not actually used for anything.

File: raftis/src/jepsen/raftis.clj [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+          :read (assoc op :type :ok, :value (parse-long (get (car/wcar conn (car/get "r")) 0 "")))
+
+          :write (do (car/wcar conn (car/set "r" (:value op)))
+                     (assoc op :type, :ok)))
+
+        (catch clojure.lang.ExceptionInfo e
+          (def err_str (str (.getMessage e)))
+          (def no_leader (re-find #"ERR write InComplete: no leader node!.*" err_str))
+          (def socket_closed (re-find #"socket closed.*" err_str))
+          (assoc op :type (if (or (or (= :read (:f op)) no_leader) socket_closed) :fail :info), :error err_str))
```
Comment (aphyr) [inline raftis/src/jepsen/raftis.clj]: You don't need to nest `or`-- `(or a b c)` works fine. In addition, I'm not sure you can treat socket closed as a definite failure.

File: raftis/src/jepsen/raftis.clj [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    (invoke! [this test op]
+      (try
+        (case (:f op)
+          :read (assoc op :type :ok, :value (parse-long (get (car/wcar conn (car/get "r")) 0 "")))
+
+          :write (do (car/wcar conn (car/set "r" (:value op)))
+                     (assoc op :type, :ok)))
+
+        (catch clojure.lang.ExceptionInfo e
+          (def err_str (str (.getMessage e)))
```
Comment (aphyr) [inline raftis/src/jepsen/raftis.clj]: `def` defines global variables, and different threads will clobber each other here. Did you mean `let`?

**Comment by gaodq** [comment]
I have fixed these bugs.

File: raftis/src/jepsen/raftis.clj [hunk 5]
```diff
@@ -41,10 +41,10 @@
                      (assoc op :type, :ok)))
 
         (catch clojure.lang.ExceptionInfo e
-          (def err_str (str (.getMessage e)))
-          (def no_leader (re-find #"ERR write InComplete: no leader node!.*" err_str))
-          (def socket_closed (re-find #"socket closed.*" err_str))
-          (assoc op :type (if (or (or (= :read (:f op)) no_leader) socket_closed) :fail :info), :error err_str))
+          (let [err_str (str (.getMessage e))]
+            (let [no_leader (re-find #"ERR write InComplete: no leader node!.*" err_str)]
+              (let [socket_closed (re-find #"socket closed.*" err_str)]
```
Comment (aphyr) [inline raftis/src/jepsen/raftis.clj]: You can provide multiple bindings in a single let: `(let [x 1, y 2 ...])`

**Comment by aphyr** [comment]
Great, thanks. :)

**Pull request merged**